src/commands/containers/files/downloadContainerFile.ts

```diff
@@ -0,0 +1,102 @@
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See LICENSE.md in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as path from 'path';
+import * as vscode from 'vscode';
+import { IActionContext, UserCancelledError } from 'vscode-azureextensionui';
+import { ext } from '../../../extensionVariables';
+import { localize } from '../../../localize';
+import { FileTreeItem } from '../../../tree/containers/files/FileTreeItem';
+import { multiSelectNodes } from '../../../utils/multiSelectNodes';
+
+async function fileExists(file: vscode.Uri): Promise<boolean> {
+    try {
+        // NOTE: The expectation is that stat() throws when the file does not exist.
+        //       Filed https://github.com/microsoft/vscode/issues/112107 to provide
+        //       a better mechanism than trapping exceptions.
+        await vscode.workspace.fs.stat(file);
+
+        return true;
+    } catch {
+        return false;
+    }
+}
+
+export async function downloadContainerFile(context: IActionContext, node?: FileTreeItem, nodes?: FileTreeItem[]): Promise<void> {
+    nodes = await multiSelectNodes(
+        { ...context, noItemFoundErrorMessage: localize('vscode-docker.commands.containers.files.downloadContainerFile.noFiles', 'No files are available to download.') },
+        ext.containersTree,
+        'containerFile',
+        node,
+        nodes
+    );
+
+    const localFolderUris = await vscode.window.showOpenDialog(
+        {
+            canSelectFiles: false,
+            canSelectFolders: true,
+            canSelectMany: false,
+            openLabel: localize('vscode-docker.commands.containers.files.downloadContainerFile.openLabel', 'Select'),
+            title: localize('vscode-docker.commands.containers.files.downloadContainerFile.openTitle', 'Select folder for download')
+        });
+
+    const localFolderUri = localFolderUris[0];
+
+    const files = nodes.map(n => {
+        const containerFileUri = n.uri;
+        const fileName = path.posix.basename(containerFileUri.path);
+
+        return {
+            containerUri: n.uri.uri,
+            fileName,
+            localUri: vscode.Uri.joinPath(localFolderUri, fileName)
+        };
+    });
+
+    await vscode.window.withProgress(
+        {
+            location: vscode.ProgressLocation.Notification,
+            title: localize('vscode-docker.commands.containers.files.downloadContainerFile.opening', 'Downloading File(s)...')
+        },
+        async (_, token) => {
+            for (const file of files) {
+                if (token.isCancellationRequested) {
+                    throw new UserCancelledError();
+                }
+
+                const localFileExists = await fileExists(file.localUri);
+
+                if (localFileExists) {
+                    const overwriteFile: vscode.MessageItem = {
```

Comment:
Maybe these messages belong to the top of the file? Or at least outside of the file loop?

Reply:
Reasonable; updated;